### ISSUE TITLE
Better stale lock detection / lease times

### ISSUE THREAD

**Issue by lissom** [restic/restic#4262]
Output of `restic version`
--------------------------
restic 0.15.1 compiled with go1.19.5 on linux/amd64
also windows

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------
Locks should have leases which timeout such that they don't hold up backups 6 hours after a restic process has died.  Such that over night processes which fail don't hold everything up until a human intervenes.

Please describe the feature you'd like us to add here.
----------------------------------------------------------------------------------
Leases on locks, such that an oom killed processes (or any other reason) doesn't halt backups.
If it needs to be a command line setting, e.g. lease-hours and the caller ignores leases older than X hours that works too.  i.e. we terminate restic processes that are running for more than an hour (all ops take only a few minutes) so we know that lock of an hour or more is definitely a dead process.  Sort of seems that if restic updates the locks every minute for a running process or something similar restic should know this too.

What are you trying to do? What problem would this solve?
---------------------------------------------------------
Not need manual intervention for processes that die.
After the oom killer on a server hitting all repos running forgot --prune, on the client machine:
Fatal: unable to create lock in backend: repository is already locked exclusively by PID 8187 on 71a1b858ce50 by $USER (UID 123, GID 456)
lock was created at 2023-03-20 03:38:21 (6h38m0.268490922s ago)

After this is implemented a single over night job going wrong shouldn't stop backups for hours.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Yes, it's a great product.

**Comment by lissom** [comment]
Also useful would be 1) list locks show timing user info and 2) unlock being able to be set to only remove a certain users locks.

**Comment by lukastribus** [comment]
A lock that expires on its own without certainty that the process is no longer running is not a proper lock and would be dangerous.

user/host/time informations in locks, displayed in `restic list locks` and to be able to use that when deleting locks in `restic unlock` sounds like a good improvement.

**Comment by MichaelEischer** [comment]
Restic refreshes locks every five minutes while it is running. Locks that are older than 30 minutes are considered as stale and will be removed by a call of `unlock`. However, as it turns out there are some corner cases in which locks are not refreshed within 30 minutes, see #4274 and #4199. These issues first have to be resolved before we can think about automatically ignoring stale locks.

It is already possible to inspect individual locks using `restic cat lock ...`.

**Comment by lissom** [comment]
It would better if all locks could be listed with details in a single command.  With cat and many locks it's hard to be sure non were missed and mistakes could lead to data loss vs all locks complete info is display and it's easy (and it's likely if something automated is going wrong, there are a lot of locks).

#4274 and #4199 getting people to not need to use 0.14 first does seem necessary.

**Comment by lissom** [comment]
It would better if all locks could be listed with details in a single command.  With cat and many locks it's hard to be sure none were missed and mistakes could lead to data loss vs all locks complete info is display and it's easy (and it's likely if something automated is going wrong, and there are a lot of locks).

#4274 and #4199 getting people to not need to use 0.14 first does seem necessary.

**Comment by MichaelEischer** [comment]
Automatically ignoring / unlocking stale locks has also been requested in https://github.com/restic/restic/issues/2736 .

**Comment by MichaelEischer** [comment]
Closing as duplicate of #2736.
